Make TraceUtils.h to be device-agnostic

fffrog opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126969

Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

fffrog committed:
Make TraceUtils.h to be device-agnostic

[ghstack-poisoned]

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h,
so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in
TraceUtils.h since I'm not sure if other devices will
use them later.

[ghstack-poisoned]

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

ezyang commented:
@kwen2501 is a more logical reviewer here, I think

FFFrog commented:
Sorry to bother you @kwen2501, could you please help take a look at this PR that is blocking the third-party backend development?

wconstab commented:
ill review this today

wconstab commented:
i think this is fine.

however i want to let https://github.com/pytorch/pytorch/pull/127421 land first and rebase this on top, just to make sure there aren't any weird land races.

FFFrog commented:
> i think this is fine.
> 
> 
> 
> however i want to let https://github.com/pytorch/pytorch/pull/127421 land first and rebase this on top, just to make sure there aren't any weird land races.

Thank you for your comment.

I will rebase this commit when https://github.com/pytorch/pytorch/pull/127421 land later and request your review again.

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

c-p-i-o commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -517,6 +519,373 @@ struct ncclRedOpRAII {
+    std::vector<c10::IValue> all_frames;
+    for (const auto& f : stracebacks.all_frames) {
+      auto d = new_dict();
+      d.insert(name_key, f.funcname);
+      d.insert(filename_key, f.filename);
+      d.insert(line_key, int64_t(f.lineno));
```
I think you can also move all the `key` definition fields out from `TraceUtils.h` and into NCCLUtils.h

FFFrog replied:
Sure, I have moved it into NCCLUtils.h.

Could you help to take a look at it again?

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

c-p-i-o commented:
Hey @FFFrog, sorry you might have to rebase and land one more time as I added a small PR:
https://github.com/pytorch/pytorch/pull/128190
that introduces a new field in the files.

c-p-i-o reviewed (approved):
Looks good to me but don't forget to rebase and integrate small change from 
https://github.com/pytorch/pytorch/pull/128190
into this change.

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

FFFrog commented:
> Looks good to me but don't forget to rebase and integrate small change from #128190 into this change.

lol, thank you a lot, i will rebase it now and re-trigger it again.

clee2000 commented:
@pytorchbot revert -m "failing internal builds D58443769" -c ghfirst

clee2000 commented:
Internal builds broke with 
```
torch/csrc/distributed/c10d/NCCLUtils.hpp:675:14: error: missing field 'duration_' initializer [-Werror,-Wmissing-field-initializers]
[CONTEXT]         isP2P};
```

@c-p-i-o Can you help provide more information?

FFFrog commented:
@c-p-i-o @clee2000 hi, sorry to bother you guy again, could you please show me more informations about this revert?

c-p-i-o commented:
> D58443769

Hey @FFFrog - maybe something didn't quite merge correctly? 
Let's sync and see what we need to do to land the changes.

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

FFFrog commented:
> > D58443769
> 
> Hey @FFFrog - maybe something didn't quite merge correctly? Let's sync and see what we need to do to land the changes.

@c-p-i-o, I've resynced and double checked all the code I've changed and don't see anything odd, could you please give me some help?

c-p-i-o commented:
> > > D58443769
> > 
> > 
> > Hey @FFFrog - maybe something didn't quite merge correctly? Let's sync and see what we need to do to land the changes.
> 
> @c-p-i-o, I've resynced and double checked all the code I've changed and don't see anything odd, could you please give me some help?

Hey @FFFrog the code is fine but it needs to be fixed up to handle some internal use cases. 
Drop me a note on chat and we can debug this further.

c-p-i-o commented:
Hey @FFFrog ... oops sorry, I thought you were an internal meta employee.  Internally, this diff failed some validation tests:
```
fbcode/caffe2/torch/csrc/distributed/c10d/NCCLUtils.hpp:675:14: error: missing field 'duration_' initializer [-Werror,-Wmissing-field-initializers]
        isP2P};
             ^
```
Let me see if I can come up with a diff to fix it.

c-p-i-o commented:
Ok figured this out. We just need to change:
```
    auto te = Entry{
        id_,
        pg_id,
        pg_name,
        collective_seq_id,
        p2p_seq_id,
        op_id,
        std::move(profiling_name),
        std::move(traceback),
        std::move(start),
        std::move(end),
        c10::getTime(),
        timeout_ms.count(),
        isP2P};
   ```
   to include additional structure fields:
   ```
       auto te = Entry{
        id_,
        pg_id,
        pg_name,
        collective_seq_id,
        p2p_seq_id,
        op_id,
        std::move(profiling_name),
        std::move(traceback),
        std::move(start),
        std::move(end),
        c10::getTime(),
        timeout_ms.count(),
        isP2P,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        {},
        {},
        {},
        {},
        {},
        false};
  ```
    Note the additional 9 fields at the bottom.
    It looks like an internal meta build is treating all warns are build fails :-/
    
    Sorry about this.

c-p-i-o commented:
@FFFrog LMK if you have any issues making the change above.

FFFrog commented:
> Ok figured this out. We just need to change:
> 
> ```
>    auto te = Entry{
>         id_,
>         pg_id,
>         pg_name,
>         collective_seq_id,
>         p2p_seq_id,
>         op_id,
>         std::move(profiling_name),
>         std::move(traceback),
>         std::move(start),
>         std::move(end),
>         c10::getTime(),
>         timeout_ms.count(),
>         isP2P};
> ```
> 
> to include additional structure fields:
> 
> ```
>        auto te = Entry{
>         id_,
>         pg_id,
>         pg_name,
>         collective_seq_id,
>         p2p_seq_id,
>         op_id,
>         std::move(profiling_name),
>         std::move(traceback),
>         std::move(start),
>         std::move(end),
>         c10::getTime(),
>         timeout_ms.count(),
>         isP2P,
>         std::nullopt,
>         std::nullopt,
>         std::nullopt,
>         {},
>         {},
>         {},
>         {},
>         {},
>         false};
> ```
> 
> Note the additional `9` fields at the bottom after `isP2P`. It looks like an internal meta build is treating all warns as build fails :-/ In this case there is an internal build that has `-Werror,-Wmissing-field-initializers` set in the build params. I don't know why this didn't fail previously but I suspect it might have been protected the `#ifdef USE_C10D_NCCL` flag. Anyway, the change above should help the internal build to pass. I tested it locally and things were good!
> 
> Sorry about this.

Thank you a lot :D, I will fix it right now by following your changes.

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

c-p-i-o commented:
LGTM!

FFFrog commented:
@pytorchbot rebase -b main

fffrog committed:
Update on "Make TraceUtils.h to be device-agnostic"


Some features of third-party devices depend on TraceUtils.h, so some of the CUDA code was removed and split into NCCLUtils files.

In addition, some common functions still remain in TraceUtils.h since I'm not sure if other devices will use them later.

[ghstack-poisoned]

FFFrog commented:
@pytorchbot merge
